app/controllers/tag_articles_controller.rb

```diff
@@ -0,0 +1,65 @@
+class TagArticlesController < ApplicationController
+  before_action :set_cache_control_headers, only: :index
+
+  SIGNED_OUT_RECORD_COUNT = 60
+
+  rescue_from ArgumentError, with: :bad_request
+
+  def index
+    @page = (params[:page] || 1).to_i
+    @article_index = true
+    handle_tag_index
+  end
+
+  private
+
+  def handle_tag_index
```

Comment:
I'm not sure we should separate this from `def index` anymore.

There are opportunities for refactoring but as it is, it should proably be one single method.

This is how I envision it:

1. load objects (@tag, @tag_model, @moderators and so on)
2. check `if @tag_model.alias_for.present?` and redirect accordingly
3. separate @num_published_aricles and @number_of_articles in its own method
4. separate the part of loading stories in its own method
5. call the two caching methods

this way you have:

```ruby
def index
  # ... assign common instance vars ...

  # if tag alias redirect

  set_number_of_articles
  set_stories

  set_surrogate_key_header
  set_cache_control_headers
end

private

def set_number_of_articles
end

def set_stories
end
```

how does that sound?

Reply:
Thanks for your advice, this is fantastic!